numba/core/types/function_type.py

```diff
@@ -110,6 +117,12 @@ def __init__(self, nargs, dispatchers):
         # make the undefined function type instance unique
         type(self)._counter += 1
         self._key += str(type(self)._counter)
+        self.instance_cache[frozenset(dispatchers)] = self
```

Comment:
The `instance_cache` is holding Dispatcher objects alive and may lead to memory leak for long running process. At the same time, it seems tricky to make it release memory automatically using a container from `weakref`.

Reply:
@sklam, I've been thinking about how can we re-use type instances without necessarily holding live references to a lot of objects. It's very similar problem as this one https://github.com/numba/numba/pull/5579#issuecomment-617371499. ie, . In there you mention using `.key` to identify repeating objects.
Alternatively,  my solution in https://github.com/luk-f-a/numba/blob/222f0fd33082bd39b18a9adec97d4a18469a2599/numba/core/types/abstract.py#L48-L74 was to use the hash (instead of the object) as key in the dictionary. 

I think either way could work here. If you tell me which one you prefer, I'll make the change.

Reply:
Using `.key` is better. The `Type.__hash__` is by default `hash(type_instance.key)`. The hash can't guarantee unique result for different `.key`

Reply:
I see why you need to use hash. Some `.key` are containing type instances.

Reply:
What about changing `key` to be weak references or creating a new `weak_key` property holding weak references or a `string_key` which is only text?